Lib/test/test_socket.py

```diff
@@ -7058,21 +7061,95 @@ def close_fds(fds):
             socket.send_fds(sock1, [MSG], fds)
             # request more data and file descriptors than expected
             msg, fds2, flags, addr = socket.recv_fds(sock2, len(MSG) * 2, len(fds) * 2)
-            self.addCleanup(close_fds, fds2)
+            self._cleanup_fds(fds2)
 
         self.assertEqual(msg, MSG)
         self.assertEqual(len(fds2), len(fds))
         self.assertEqual(flags, 0)
         # don't test addr
 
         # test that file descriptors are connected
-        for index, fds in enumerate(pipes):
-            rfd, wfd = fds
-            os.write(wfd, str(index).encode())
+        for index, ((_, wfd), rfd) in enumerate(zip(pipes, fds2)):
+            self._test_pipe(rfd, wfd, str(index).encode())
+
+    @unittest.skipUnless(sys.platform in ("linux", "android", "darwin"),
+                         "works on Linux and macOS")
+    def test_send_recv_fds_with_addrs(self):
+        sock1 = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
+        sock2 = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
+        rfd, wfd = os.pipe()
+        self.addCleanup(os.close, rfd)
+        self.addCleanup(os.close, wfd)
+
+        with tempfile.TemporaryDirectory() as tmpdir, sock1, sock2:
+            sock1_addr = os.path.join(tmpdir, "sock1")
+            sock2_addr = os.path.join(tmpdir, "sock2")
+            sock1.bind(sock1_addr)
+            sock2.bind(sock2_addr)
+            sock2.setblocking(False)
+
+            socket.send_fds(sock1, [MSG], [rfd], address=sock2_addr)
+            msg, fds, flags, addr = socket.recv_fds(sock2, len(MSG), 1)
+            self._cleanup_fds(fds)
+
+        self.assertEqual(msg, MSG)
+        self.assertEqual(len(fds), 1)
+        self.assertEqual(addr, sock1_addr)
+
+        self._test_pipe(fds[0], wfd, MSG)
 
-        for index, rfd in enumerate(fds2):
-            data = os.read(rfd, 100)
-            self.assertEqual(data,  str(index).encode())
+    @requireAttrs(socket, "MSG_PEEK")
+    @unittest.skipUnless(sys.platform in ("linux", "android"), "works on Linux")
```

Comment:
Is there a way to have a flag test on macOS or not at all?

Reply:
I thought about it, but I'm not sure how to implement it. Is it good practice to write a `_test_recv_fds_peek(self, skip_fd_check=False)` method and 2 wrappers with different arguments?

Reply:
A good practice is to check if there's a flag that is supported both on Unix and Windows. The reason is that the availability of `recv_fds` is marked as Unix and Windows but not WASI. So we shouldn't have this kind of skip in the first place.

Now, if there is not a common flag that is supported on all platforms, just writing three different tests for three different flags is also fine (each of which protected by a `skipUnless`)

Reply:
> The reason is that the availability of recv_fds is marked as Unix and Windows but not WASI. So we shouldn't have this kind of skip in the first place.

`recv_fds` requires `recvmsg`, which is only available on Unix but not WASI, so I think it may be a documentation error.

Reply:
Ok actually we have https://github.com/python/cpython/issues/122153 which wants to implement it but we never implemented it. Good to know that it's a doc issue. We can open a PR (using 122153 as the GH issue number) to amend the docs. For now, can you add some comment saying that it's not available on Windows yet but can be (and add some "see gh-122153)? thanks

Reply:
> For now, can you add some comment saying that it's not available on Windows yet but can be (and add some "see [gh-122153](https://github.com/python/cpython/issues/122153))? thanks

Of course!

However sendmsg/recvmsg aren't the only requirements, CPython doesn't support `AF_UNIX` on Windows, and Windows doesn't support `SCM_RIGHTS` at all, so I think it's not possible to implement send_fds/recv_fds on Windows.

ref: https://github.com/python/cpython/issues/77589#issuecomment-1963245972

Reply:
This doc issue was introduced in e3b6ff19aaa318a813130ba9ad2ab0a332f27feb